STYLEGUIDE.md

```diff
@@ -0,0 +1,381 @@
+letter of each word except the first one is capitalized, prefer to capitalize
+abbreviations as single words, e.g., startRpc() rather than startRPC().
+
+Template parameters should follow the naming style for their category: type
+template parameters should follow the rules for type names, and non-type
+template parameters should follow the rules for variable names.
+
+### File Names
+
+Filenames start with a capital letter and have a capital letter for each new
+word, with no underscores: `MyUsefulClass.cpp`, `VeryUsefulClass.cpp`
+
+C++ files should end in `.cpp` and header files should end in `.h`. Files that
+rely on being textually included at specific points should end in `.inc` (see
+also the section on self-contained headers). Do not use filenames that already
+exist in `/usr/include`, such as `db.h`. In general, make your filenames very
+specific. For example, use `HttpServerLogs.h` rather than `Logs.h`. A very
+common case is to have a pair of files called, e.g., `FooBar.h` and
+`FooBar.cpp`, defining a class called `FooBar`.
+
+### Type Names
+
+Type names start with a capital letter and have a capital letter for each new
+word, with no underscores: `MyExcitingClass`, `MyExcitingEnum`.
+
+The names of all types — classes, structs, type aliases, enums, and type
+template parameters — have the same naming convention. Type names should start
+with a capital letter and have a capital letter for each new word. No
+underscores. For example:
+
+```
+// classes and structs
+class UrlTable { ...
+class UrlTableTester { ...
+struct UrlTableProperties { ...
+
+// typedefs
+typedef hash_map<UrlTableProperties*, std::string> PropertiesMap;
+
+// using aliases
+using PropertiesMap = hash_map<UrlTableProperties*, std::string>;
+
+// enums
+enum class UrlTableError { ...
+```
+
+### Variable Names
+
+The names of variables (including function parameters) should follow "[camel
+case](https://en.wikipedia.org/wiki/Camel_case)" practice with no underscores
+between words. Data members of classes (but not structs) additionally have
+leading underscores. For instance: `aLocalVariable`, `aStructDataMember`,
+`_aClassDataMember`.
+
+#### Common Variable names
+
+For example:
+
+```
+std::string table_name;  // BAD - lowercase with underscore.
+
+std::string tableName;   // Ok - mixed case.
+```
+
+#### Class Data Members
+
+Data members of classes, both static and non-static, are named like ordinary
+nonmember variables, but with a leading underscore.
+
+```
+class TableInfo {
+  ...
+ private:
+  std::string _tableName;  // OK - mixed case with leading underscore
+  static Pool<TableInfo>* _pool;  // OK.
+};
+```
+
+#### Struct Data Members
+
+Data members of structs, both static and non-static, are named like ordinary
+nonmember variables. They do not have the leading underscores that data members
+in classes have.
+
+```
+struct UrlTableProperties {
+  std::string name;
+  int numEntries;
+  static Pool<UrlTableProperties>* pool;
+};
+```
```

Comment:
Could we please change this to talk about public/private data members instead of struct/class data members, because that's actually what this is about.
```
class UrlTableProperties {
public:
  std::string name;
  int numEntries;
  static Pool<UrlTableProperties>* pool;
};
```
Is the same as the struct, yet according to the current wording this one should use underscores.

Also, it is possible to have both, private and public data members in the same class, e.g.:
```
class TableInfo {
  ...
  std::size_t const size;
 private:
  std::string _tableName;
  static Pool<TableInfo>* _pool;
};
```
In this case I would not want to prefix the publicly visible data member, but only the private ones.